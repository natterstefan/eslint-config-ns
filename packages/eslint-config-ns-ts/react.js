/**
 * @type {import('eslint').ESLint.Options}
 */
module.exports = {
  extends: ['./rules/react'].map(require.resolve),
  parser: '@typescript-eslint/parser',
  rules: {},
}
