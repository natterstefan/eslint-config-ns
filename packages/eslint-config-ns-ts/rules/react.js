/**
 * @type {import('eslint').ESLint.Options}
 */
module.exports = {
  rules: {
    // https://www.npmjs.com/package/eslint-plugin-react
    // WHY? Because we focus on interfaces and types instead
    'react/prop-types': 0,
    'react/require-default-props': 0,
  },
}
