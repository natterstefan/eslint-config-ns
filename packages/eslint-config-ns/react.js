/**
 * @type {import('eslint').ESLint.Options}
 */
module.exports = {
  extends: ['./rules/react', './rules/react-a11y'].map(require.resolve),
  parser: '@babel/eslint-parser',
  parserOptions: {
    babelOptions: {
      presets: ['@babel/preset-react'],
    },
    ecmaFeatures: {
      jsx: true,
    },
    requireConfigFile: false,
  },
  globals: {
    __DEV__: true,
  },
  rules: {},
}
