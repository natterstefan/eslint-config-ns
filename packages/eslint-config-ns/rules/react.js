/**
 * @type {import('eslint').ESLint.Options}
 */
module.exports = {
  extends: [
    'airbnb/hooks',
    // https://www.npmjs.com/package/eslint-plugin-react
    'plugin:react/recommended',
    // https://www.npmjs.com/package/eslint-plugin-react-hooks
    'plugin:react-hooks/recommended',
  ],
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
  },
  rules: {
    // https://www.npmjs.com/package/eslint-plugin-react-hooks
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'error',

    // https://www.npmjs.com/package/eslint-plugin-react
    // use "" when passing a string as a property
    'react/jsx-curly-brace-presence': [
      2,
      { props: 'never', children: 'never' },
    ],
    // jsx is also allowed in non .jsx files
    'react/jsx-filename-extension': 0,
    // Allowed but we should be aware to not overuse this.
    'react/jsx-props-no-spreading': 0,
    // Enforce component methods order
    'react/sort-comp': 2,
    // Enforce a specific function type for function components
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md
    'react/function-component-definition': [
      2,
      {
        namedComponents: ['function-declaration', 'arrow-function'],
        unnamedComponents: ['function-expression', 'arrow-function'],
      },
    ],
  },
  settings: {
    react: {
      // Tells eslint-plugin-react to automatically detect the version of React
      // to use
      version: 'detect',
    },
  },
}
