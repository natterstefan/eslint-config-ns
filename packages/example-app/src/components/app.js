import React, { useState } from 'react'

import { Text } from './text'

export const App = () => {
  const [counter, setCounter] = useState(0)
  const headline = 'Hello eslint-config-ns'

  setInterval(() => setCounter(counter + 1), 1000)

  return (
    <>
      <h1>{headline}</h1>
      <Text />
      <p>State Counter: {counter}</p>
    </>
  )
}
