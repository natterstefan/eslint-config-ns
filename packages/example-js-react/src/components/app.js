import React, { useEffect, useState } from 'react'

import { Text } from './text'

export const App = () => {
  const [counter, setCounter] = useState(0)
  const headline = 'Hello eslint-config-ns'

  useEffect(() => {
    setInterval(() => setCounter(c => c + 1), 1000)
  }, [])

  return (
    <>
      <h1>{headline}</h1>
      <Text text="Hello World" />
      <p>State Counter: {counter}</p>
    </>
  )
}
