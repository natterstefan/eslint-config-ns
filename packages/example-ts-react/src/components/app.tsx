import React, { useState, FunctionComponent } from 'react'

import { Text } from './text'

export const App: FunctionComponent = () => {
  const [counter, setCounter] = useState(0)
  const headline = 'Hello eslint-config-ns-ts'

  setInterval(() => setCounter(counter + 1), 1000)

  return (
    <>
      <h1>{headline}</h1>
      <Text text="Hello World" />
      <p>State Counter: {counter}</p>
    </>
  )
}
