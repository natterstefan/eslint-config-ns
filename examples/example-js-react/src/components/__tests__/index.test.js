import React from 'react'
import { act } from 'react-dom/test-utils'
import { render, screen } from '@testing-library/react'

import { App } from '../app'

jest.useFakeTimers()

describe('App', () => {
  it('renders', () => {
    const { container } = render(<App />)
    expect(container.innerHTML).toMatchSnapshot()
  })

  it('increases counter every 1 second', () => {
    render(<App />)
    const elem = screen.getByTestId('counter')
    expect(elem.innerHTML).toBe('0')

    act(() => {
      jest.advanceTimersByTime(1000)
    })

    expect(elem.innerHTML).toBe('1')

    act(() => {
      jest.advanceTimersByTime(1000)
    })
    expect(elem.innerHTML).toBe('2')
  })
})
