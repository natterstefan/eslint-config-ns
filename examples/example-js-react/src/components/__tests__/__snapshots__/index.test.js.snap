// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `"<h1>Hello eslint-config-ns</h1><div>Hello World</div><p>State Counter: <span data-testid="counter">0</span></p>"`;
